Firestore/core/src/firebase/firestore/core/transaction_runner.mm

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/core/transaction_runner.h"
+#include "Firestore/core/src/firebase/firestore/remote/exponential_backoff.h"
+
+#include "absl/algorithm/container.h"
+
+using firebase::firestore::core::TransactionResultCallback;
+using firebase::firestore::core::TransactionUpdateCallback;
+using firebase::firestore::remote::Datastore;
+using firebase::firestore::remote::RemoteStore;
+using firebase::firestore::util::AsyncQueue;
+using firebase::firestore::util::TimerId;
+using firebase::firestore::util::Status;
+
+namespace firebase {
+namespace firestore {
+namespace core {
+TransactionRunner::TransactionRunner(const std::shared_ptr<AsyncQueue>& queue,
+                                     RemoteStore* remote_store,
+                                     TransactionUpdateCallback update_callback,
+                                     TransactionResultCallback result_callback)
+    : queue_{queue},
+      remote_store_{remote_store},
+      update_callback_{update_callback},
+      result_callback_{result_callback},
+      backoff_{queue_, TimerId::RetryTransaction} {
+  retries_left_ = kRetryCount;
+}
+
+void TransactionRunner::Run() {
+  queue_->VerifyIsCurrentQueue();
+
+  auto shared_this = this->shared_from_this();
+  backoff_.BackoffAndRun([shared_this] {
+    std::shared_ptr<Transaction> transaction =
+        shared_this->remote_store_->CreateTransaction();
+    shared_this->update_callback_(
+        transaction,
+        [transaction, shared_this](util::StatusOr<absl::any> maybe_result) {
+          shared_this->queue_->Enqueue(
+              [transaction, shared_this, maybe_result] {
+                shared_this->RunTransaction(transaction, maybe_result);
+              });
+        });
+  });
+}
+
+void TransactionRunner::RunTransaction(
```

Comment:
Hrm...  Splitting Run() into two methods is definitely a worthwhile goal, since it'd be pretty long / deeply nested otherwise.  But having `Run()` and `RunTransaction()` isn't super clear to me and it's a bit hard to come up with a great name.  My first thought was `CommitTransaction()` except it only commits if maybe_result is successful.  Maybe `HandleUpdateCallbackResult()`?  Or `CheckUpdateCallbackResult()` to match `CheckCommitResult()`?  WDYT?

Reply:
I see what you mean. I think `CheckUpdateCallbackResult()` makes the most sense, since the transaction is run in two steps, and this way each step has its own corresponding function name.